bokehjs/src/lib/models/ranges/factor_range.ts

```diff
@@ -97,14 +109,54 @@ export function map_three_levels(
   let total_subpad = 0
   for (const [f0, top] of tops) {
     const n = top.length
-    const [submap, subpad] = map_two_levels(top, inner_pad, factor_pad, suboffset)
-    total_subpad += subpad
-    const subtot = sum(top.map(([f1]) => submap.get(f1)!.value))
-    mapping.set(f0, {value: subtot/n, mapping: submap})
-    suboffset += n + outer_pad + subpad
+    const sub = map_two_levels(top, inner_pad, factor_pad, suboffset)
+    total_subpad += sub.inner_padding
+    const subtot = sum(top.map(([f1]) => sub.mapping.get(f1)!.value))
+    mapping.set(f0, {value: subtot/n, mapping: sub.mapping as L2Mapping})
+    suboffset += n + outer_pad + sub.inner_padding
+  }
+
+  const mids: L2Factor[] = []
+  for (const [f0, L2] of mapping) {
+    for (const f1 of L2.mapping.keys()) {
+      mids.push([f0, f1])
+    }
+  }
+
+  const inner_padding = (tops.size - 1)*outer_pad + total_subpad
+  return {mapping, tops: [...mapping.keys()], mids, inner_padding}
+}
+
+const is_l1 = isString
+const is_l2 = (x: any) => isArray(x) && x.length == 2 && isString(x[0]) && isString(x[1])
+const is_l3 = (x: any) => isArray(x) && x.length == 3 && isString(x[0]) && isString(x[1]) && isString(x[2])
```

Comment:
No more `any` type. Use `unknown`, unless you're in a situation that you have to coerce the type system:
```suggestion
const is_l2 = (x: unknown) => isArray(x) && x.length == 2 && isString(x[0]) && isString(x[1])
const is_l3 = (x: unknown) => isArray(x) && x.length == 3 && isString(x[0]) && isString(x[1]) && isString(x[2])
```
Currently I can't block `any` type during linting, but as soon as that's possible, it will go on the banned type list.